aten/src/ATen/native/quantized/cpu/qbatch_norm.cpp

```diff
@@ -39,6 +39,105 @@ void compute_fused_params(
   }
 }
 
+template <bool ReluFused>
+Tensor q_batch_norm1d_impl(
+    Tensor qx,
+    c10::optional<Tensor> mb_weight,
+    c10::optional<Tensor> mb_bias,
+    Tensor mean,
+    Tensor var,
+    double eps,
+    double output_scale,
+    int64_t output_zero_point) {
+
+  TORCH_CHECK(mb_weight.has_value(), "Weight must be provided");
+  TORCH_CHECK(mb_bias.has_value(), "Bias must be provided");
+  const auto& weight = *mb_weight;
+  const auto& bias = *mb_bias;
+
+  if (qx.numel() == 0) {
+    auto out = qx.clone();
+    return out;
+  }
+  int64_t ndim = qx.dim();
+  TORCH_CHECK(ndim == 3, "Expecting the input tensor of rank 3.");
+  const int64_t N = qx.size(0);
+  const int64_t C = qx.size(1);
+  const int64_t H = qx.size(2);
+
+  TORCH_CHECK(weight.numel() == C, "Expect weight size to match C");
+  TORCH_CHECK(bias.numel() == C, "Expect weight size to match C");
+
+  const float* weight_data = weight.template data_ptr<float>();
+  const float* bias_data = bias.template data_ptr<float>();
+
+  TORCH_CHECK(mean.numel() == C, "Mean size must match channel dimension");
+  TORCH_CHECK(var.numel() == C, "Variance size must match channel dimension");
+
+  Tensor alpha = at::empty_like(mean, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  Tensor beta = at::empty_like(mean, LEGACY_CONTIGUOUS_MEMORY_FORMAT);
+  float* alpha_data = alpha.data_ptr<float>();
+  float* beta_data = beta.data_ptr<float>();
+
+  const float* mean_data = mean.template data_ptr<float>();
+  const float* var_data = var.template data_ptr<float>();
+
+  // create a fake W dimension so we can use NHWC
+  qx = qx.unsqueeze(-1);
+
+  auto oSizes = qx.sizes();
+  auto qx_nhwc = qx.contiguous(MemoryFormat::ChannelsLast);
+  Tensor qy = at::_empty_affine_quantized(
+      oSizes,
+      at::device(kCPU)
+        .dtype(qx_nhwc.scalar_type())
+        .memory_format(MemoryFormat::ChannelsLast),
+      output_scale,
+      output_zero_point,
+      c10::nullopt);
+
+  compute_fused_params(
+      C,
+      weight_data,
+      bias_data,
+      mean_data,
+      var_data,
+      eps,
+      qx.q_scale(),
+      output_scale,
+      alpha_data,
+      beta_data);
+  if (ReluFused) {
+    qbatch_norm_relu_stub(
+        qx.device().type(),
+        N,
+        C,
+        H,
+        qx.q_zero_point(),
+        output_zero_point,
+        qx_nhwc,
+        alpha,
+        beta,
+        qy);
+  } else {
+    qbatch_norm_stub(
+        qx.device().type(),
+        N,
+        C,
+        H,
+        qx.q_zero_point(),
+        output_zero_point,
+        qx_nhwc,
+        alpha,
+        beta,
+        qy);
+  }
+  // Remove the fake dimension, and go back to contiguous format
+  // (since there is no 4th channel). Note, this has a performance
+  // cost.
+  return qy.contiguous(MemoryFormat::Contiguous).squeeze(-1);
```

Comment:
checked with Vitaly, we can add `MemoryFormat::ChannelsLast1d` as a future PR if we need to optimize this further.

Reply:
does squeeze affect the `ChannelsLast1d`? i.e. can squeeze preserve the channel order

Reply:
sorry, not 100% sure of the question, are you asking if it would work correctly if we squeeze without converting back to contiguous, and would it remove the correct dim?

Reply:
I mean does this implementation affect the memory layout of the Tensor.
let's say we implement the qbatch_norm_1d natively without squeeze, I assume that can preserve the memory layout of the Tensor, that is given a Tensor with non-contiguous NHWC layout the output would still be NHWC. and does using squeeze here mean that we need to make the Tensor contiguous?

Reply:
ah, sure.  Here is my current understanding:
* the cpu intrinsics used to make the kernel fast require the input data for a channel to be contiguous (because they do SIMD reads on elements which are adjacent in memory).  I'd guess that trying to write a kernel for BN without requiring channels to be contiguous would be slower than what we currently have.  Didn't try it though.
* the current approach is a hack to reuse the 2d kernel (add a dim, do kernel in NHWC, remove the dim).  The only reason we convert back from NHWC to contiguous at the end is because "3 channel channels-last" is not defined yet.  The data would be in the right format (and tests on this diff actually pass with 3-channel NHWC), but the scaffolding just isn't there for this data format to be read correctly by whoever will use this tensor.  
* I got an estimate of "1-2 eng weeks, high risk" to make the new memory format a first class citized from vfedunin, so we can choose to take that path in the future if the customer team really asks for it (and if this becomes a perf bottleneck)

sorry, still not 100% I answered your question - lmk if the ^ doesn't cover it

Reply:
Thanks for the clarification. I think that answers my question. My question is mostly that let's say all the previous ops are working with NHWC order and all the ops after this batchnorm is working with NHWC order, does this batchnorm break this? I guess the answer is yes, in this case. So basically here we need to go from NHWC to NCHW and then in the next op we'll go from NCHW to NHWC.

Reply:
ah, I see.  Sorry, now I fully understand the question.  I believe the way it's supposed to work is
* each op has to be able to work with either NHWC or NCHW input, so as long as the tensor is properly tagged with memory format then it's guaranteed a correct result.  If it doesn't support some input, it should throw an error.
* currently there are manual optimizations to make defaults be sane, and I think those manual guidelines ask to use NHWC if possible (in this case it's not possible yet because we need ChannelsLast1d first).
* in the future there might be graph level optimizations (i.e. converting nhwc_kernel -> nhwc_to_nchw -> ncwh_to_nhwc -> nhwc_kernel to just be nhwc_kernel -> nhwc_kernel).  I don't think that exists yet.

cc @kimishpatel  since he might know a more updated version of the guidelines on this.

Reply:
btw, for batchnorm1d specifically, there shouldn't be nhwc before and after, because of the same issue that `MemoryFormat::ChannelsLast1d` does not exist yet.